File: src/dnd_initiative_tracker/initiative.py
from __future__ import annotations

import random

from dnd_initiative_tracker.models import Combatant


def calculate_initiative_modifier(dexterity_score: int) -> int:
    return (dexterity_score - 10) // 2


def resolve_initiative_bonus(
    dexterity_score: int | None,
    explicit_bonus: int | None,
) -> int:
    if explicit_bonus is not None:
        return explicit_bonus
    if dexterity_score is None:
        return 0
    return calculate_initiative_modifier(dexterity_score)


def roll_d20(random_generator: random.Random | None = None) -> int:
    effective_random_generator = random_generator or random.Random()
    return effective_random_generator.randint(1, 20)


def assign_npc_initiative(
    combatant: Combatant,
    random_generator: random.Random | None = None,
) -> Combatant:
    initiative_roll = roll_d20(random_generator)
    initiative_bonus = resolve_initiative_bonus(
        dexterity_score=combatant.dex,
        explicit_bonus=combatant.initiative_bonus,
    )
    combatant.initiative_bonus = initiative_bonus
    combatant.initiative_roll = initiative_roll
    combatant.initiative_total = initiative_roll + initiative_bonus
    return combatant


def assign_player_initiative(combatant: Combatant, initiative_roll: int) -> Combatant:
    initiative_bonus = resolve_initiative_bonus(
        dexterity_score=combatant.dex,
        explicit_bonus=combatant.initiative_bonus,
    )
    combatant.initiative_bonus = initiative_bonus
    combatant.initiative_roll = initiative_roll
    combatant.initiative_total = initiative_roll + initiative_bonus
    return combatant


def sort_combatants_for_initiative(combatants: list[Combatant]) -> list[Combatant]:
    sorted_combatants = sorted(
        combatants,
        key=lambda combatant: (
            combatant.initiative_total if combatant.initiative_total is not None else -1,
            combatant.initiative_bonus if combatant.initiative_bonus is not None else -99,
            -combatant.sort_index,
        ),
        reverse=True,
    )
    for sort_index, combatant in enumerate(sorted_combatants):
        combatant.sort_index = sort_index
    return sorted_combatants

File: internal/tracker/initiative.go
package tracker

import (
	"math/rand/v2"
	"sort"
)

func CalculateInitiativeModifier(dexterityScore int) int {
	return (dexterityScore - 10) / 2
}

func ResolveInitiativeBonus(dexterityScore *int, explicitBonus *int) int {
	if explicitBonus != nil {
		return *explicitBonus
	}
	if dexterityScore == nil {
		return 0
	}
	return CalculateInitiativeModifier(*dexterityScore)
}

func RollD20() int {
	return rand.IntN(20) + 1
}

func AssignNPCInitiative(combatant *Combatant) {
	roll := RollD20()
	bonus := ResolveInitiativeBonus(combatant.Dex, combatant.InitiativeBonus)
	combatant.InitiativeBonus = &bonus
	combatant.InitiativeRoll = &roll
	total := roll + bonus
	combatant.InitiativeTotal = &total
}

func AssignPlayerInitiative(combatant *Combatant, initiativeRoll int) {
	bonus := ResolveInitiativeBonus(combatant.Dex, combatant.InitiativeBonus)
	combatant.InitiativeBonus = &bonus
	combatant.InitiativeRoll = &initiativeRoll
	total := initiativeRoll + bonus
	combatant.InitiativeTotal = &total
}

func SortCombatantsForInitiative(combatants []Combatant) []Combatant {
	sortedCombatants := cloneCombatants(combatants)
	sort.Slice(sortedCombatants, func(left, right int) bool {
		leftCombatant := sortedCombatants[left]
		rightCombatant := sortedCombatants[right]

		leftTotal := -1
		rightTotal := -1
		if leftCombatant.InitiativeTotal != nil {
			leftTotal = *leftCombatant.InitiativeTotal
		}
		if rightCombatant.InitiativeTotal != nil {
			rightTotal = *rightCombatant.InitiativeTotal
		}
		if leftTotal != rightTotal {
			return leftTotal > rightTotal
		}

		leftBonus := -99
		rightBonus := -99
		if leftCombatant.InitiativeBonus != nil {
			leftBonus = *leftCombatant.InitiativeBonus
		}
		if rightCombatant.InitiativeBonus != nil {
			rightBonus = *rightCombatant.InitiativeBonus
		}
		if leftBonus != rightBonus {
			return leftBonus > rightBonus
		}

		return leftCombatant.SortIndex < rightCombatant.SortIndex
	})

	for index := range sortedCombatants {
		sortedCombatants[index].SortIndex = index
	}

	return sortedCombatants
}
